app/src/main/java/org/mozilla/fenix/FenixApplication.kt

```diff
@@ -65,6 +66,12 @@ open class FenixApplication : Application() {
         // on the main process, as it uses Gecko to fetch experiments from the server.
         experimentLoader = loadExperiments()
 
+        // Enable the service-experiments component
+        Experiments.initialize(
+            applicationContext,
+            mozilla.components.service.experiments.Configuration(httpClient = lazy { components.core.client })
```

Comment:
Just a heads-up that we have a potential problem here: `components.core.client` is using the Gecko fetch client and could be the first thing that needs a Gecko runtime. The runtime must be initialized on the main thread and this change could potentially cause initialization of the lazy `httpClient` (and therefore the Gecko runtime) on a background thread. This would happen if `service-experiments` is the first thing needing a Gecko runtime (which is likely as the Home fragment doesn't need one).

In `service-fretboard` the field isn't lazy and the httpClient is therefore initialized on the main thread: https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L139

TL;DR: I think we need a change in Fenix to make sure the Gecko runtime is only ever created on the main thread before this change lands.
//cc @boek @pocmo @hawkinsw (as we've talked about this before)

Reply:
👍 https://github.com/mozilla-mobile/android-components/issues/3741

Reply:
@csadilek thanks for the heads up!  So would it be safer to just remove the lazy instantiation here or do I need to wait until this issue is resolved?

Reply:
Also, I have other reasons to wait on this PR until https://github.com/mozilla-mobile/android-components/pull/3821 lands in A-C and is available in Fenix.

Reply:
@travis79 I think this is something we have to fix/guarantee in Fenix either way. However, there's another race condition in the `service-fretboard` integration and I was curious how that's best addressed using `service-experiments`. Experiments are loaded from storage and fetched from the network on startup (both on IO threads), but other components / parts of the app will check `isInExperiment` on the main thread (this currently happens on startup as well). There's no guarantee that `isInExperiment` will be called after the experiments have been read from storage: https://github.com/mozilla-mobile/fenix/blob/5ecdce6ba6c121cf15d915c12b52803cf564be95/app/src/main/java/org/mozilla/fenix/components/BackgroundServices.kt#L73

The way I see it, we need a mechanism to either block on experiments being loaded from storage (not network) or some other callback solution? @hawkinsw provided a solution that isn't currently used though: https://github.com/mozilla-mobile/fenix/blob/5ecdce6ba6c121cf15d915c12b52803cf564be95/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L97 

We might be using this incorrectly though. Wdyt?

Reply:
I see what you mean, you don't really want to block in `onCreate()` so it's calling the `loadExperiments()` function of Fretboard on the IO thread.  

I think service-experiments may suffer from the same problem as it does require a little startup to read from storage, the difference being that service-experiments only stores the active experiment rather than the entire `experimentsResult` as retrieved from the server, and it stores it in SharedPreferences rather than a flat file.  The other difference is that service-experiments is using WorkManager to do the background work of fetching from the server so it shouldn't be slowing down the main thread. Once the WorkManager Worker receives the response from the server, it updates the active experiment info, but it doesn't currently have a mechanism to notify the application.

Being able to load this asynchronously is definitely going to lower startup impact and that could require some way to synchronize on loading the experiments before we go checking to see if we are in an experiment/branch.  I'm not sure if it's too much to ask of the experiments library or fretboard to handle this, or if it's more flexible to deal with it in the app like @hawkinsw's solution.

Personally, I like @hawkinsw's solution as it gives the app more control over how this will be executed in the background and awaited.

Reply:
I filed [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1569647) to follow up on the callback question.